geoutils/raster/interpolate.py

```diff
@@ -0,0 +1,329 @@
+
+method_to_order = {"nearest": 0, "linear": 1, "cubic": 3, "quintic": 5, "slinear": 1, "pchip": 3, "splinef2d": 3}
+
+
+def _dist_nodata_spread(order: int, dist_nodata_spread: Literal["half_order_up", "half_order_down"] | int) -> int:
+    """
+    Derive distance of nodata spreading based on interpolation order.
+
+    :param order: Interpolation order.
+    :param dist_nodata_spread: Spreading distance of nodata, either half-order rounded up (default), rounded down, or
+        fixed integer.
+    """
+
+    if dist_nodata_spread == "half_order_up":
+        dist_nodata_spread = int(np.ceil(order / 2))
+    elif dist_nodata_spread == "half_order_down":
+        dist_nodata_spread = int(np.floor(order / 2))
+
+    return dist_nodata_spread
+
+
+def _interpn_interpolator(
+    coords: tuple[NDArrayNum, NDArrayNum],
+    values: NDArrayNum,
+    fill_value: Number = np.nan,
+    bounds_error: bool = False,
+    dist_nodata_spread: Literal["half_order_up", "half_order_down"] | int = "half_order_up",
+    method: Literal["nearest", "linear", "cubic", "quintic", "slinear", "pchip", "splinef2d"] = "linear",
+) -> Callable[[tuple[NDArrayNum, NDArrayNum]], NDArrayNum]:
+    """
+    Create SciPy interpolator with nodata spreading. Default is spreading at distance of half the method order
+    rounded up (i.e., linear spreads 1 nodata in each direction, cubic spreads 2, quintic 3).
+
+    Gives the exact same result as scipy.interpolate.interpn, and allows interpolator to be re-used if required (
+    for speed).
+    In practice, returns either a NaN-modified RegularGridInterpolator or a NaN-modified RectBivariateSpline object,
+    both expecting a tuple of X/Y coordinates to be evaluated.
+
+    Adapted from:
+    https://github.com/scipy/scipy/blob/44e4ebaac992fde33f04638b99629d23973cb9b2/scipy/interpolate/_rgi.py#L743.
+    """
+
+    # REMOVED (native NaN support not reliable)
+    # Adding masking of NaNs for methods not supporting it
+    # method_support_nan = method in ["nearest", "linear"]
+
+    # Derive distance to spread nodata to depending on method order
+    order = method_to_order[method]
+    d = _dist_nodata_spread(order=order, dist_nodata_spread=dist_nodata_spread)
+
+    # If NaNs are not supported
+    if True:
+        # We compute the mask and dilate it to the distance to spread nodatas
+        mask_nan = ~np.isfinite(values)
+        if d != 0:
+            new_mask = binary_dilation(mask_nan, iterations=d).astype("uint8")
+        # Zero iterations has a different behaviour in binary_dilation than doing nothing, we want the original array
+        else:
+            new_mask = mask_nan.astype("uint8")
+
+        # We create an interpolator for the mask using nearest
+        interp_mask = RegularGridInterpolator(
+            coords, new_mask, method="nearest", bounds_error=bounds_error, fill_value=1
+        )
+
+        # We replace all NaN values by nearest neighbours to minimize interpolation errors near NaNs
```

Comment:
Since interpolated values are exactly at a pixel, which neighbor is selected if all 4 closest neighbors are valid? 
Why not do a bilinear interpolation, is it too costly?

Reply:
You mean in the case where there is a NaN surrounded by 4 valid pixels? It's true that this filling method would attribute any nearest value then... But the NaN propagation would also yield a NaN everywhere in the surrounding of this pixel, so it wouldn't really create an issue.

The idea in this bit of code here is just to avoid giving a very far-off value (let's say 0 or -9999) compare to the raster value in this area (let's say 8000 if it's a DEM near Everest) during cubic/quintic interpolation, as these interpolations do not support NaNs at all and requires to fill the raster with values everywhere. By feeding the nearest value, we give a value within the same order of magnitude, and so we don't create any effects near the edges.

We're talking about avoiding errors of 10e-2/e-3 relative to the signal.
I'm adding a comment to clarify this in the code!

Reply:
And for the default nodata propagation method (half order up), this is also not too much of an issue as this nearest value should be largely out of the interpolation kernel, but it mitigates errors for shorter distance of nodata propagation :)